File: src/gwcomp/options.cpp
#include "gwcomp/options.hpp"

#include "config.hpp"

#include <algorithm>
#include <charconv>
#include <filesystem>
#include <limits>
#include <ostream>
#include <string_view>
#include <utility>

namespace glasswyrm::compositor {
namespace {

void print_usage(std::ostream& output) {
  output <<
      "Usage: gwcomp [--backend headless|drm] --ipc-socket PATH\n"
      "  headless: --dump-dir PATH [--headless-output "
      "NAME[:WIDTHxHEIGHT[@MILLIHZ]]]...\n"
      "            [--headless-vrr NAME=MIN-MILLIHZ-MAX-MILLIHZ]...\n"
      "            [--scene-manifest PATH]\n"
      "  drm direct: --drm-device PATH|auto --tty /dev/ttyN\n"
      "  drm external: --drm-fd N --external-session\n"
      "  drm options: [--connector NAME] [--mode WIDTHxHEIGHT[@MILLIHZ]]\n"
      "               [--drm-api auto|atomic|legacy]\n"
      "               [--mirror-dump-dir PATH]\n"
      "               [--mirror-dump-trigger PATH] [--drm-report PATH]\n"
      "  renderer: [--renderer software|gles|auto] [--renderer-report PATH]\n"
      "  common: [--vrr-report PATH] [--once] [--max-frames N]\n"
      "          [--help] [--version]\n";
}

bool parse_positive(std::string_view text, std::uint64_t& value) {
  if (text.empty()) return false;
  const auto [end, error] =
      std::from_chars(text.data(), text.data() + text.size(), value);
  return error == std::errc{} && end == text.data() + text.size() && value > 0;
}

bool take_path(int argc, char** argv, int& index, std::string& destination,
               std::string_view option, std::ostream& error) {
  if (++index >= argc || argv[index][0] == '\0') {
    error << "gwcomp: " << option << " requires a non-empty path\n";
    return false;
  }
  destination = argv[index];
  return true;
}

bool parse_nonnegative_fd(std::string_view text, int& value) {
  if (text.empty()) return false;
  unsigned parsed = 0;
  const auto [end, error] =
      std::from_chars(text.data(), text.data() + text.size(), parsed);
  if (error != std::errc{} || end != text.data() + text.size() ||
      parsed > static_cast<unsigned>(std::numeric_limits<int>::max()))
    return false;
  value = static_cast<int>(parsed);
  return true;
}

bool parse_dimension(std::string_view text, std::uint32_t& value) {
  if (text.empty()) return false;
  const auto [end, error] =
      std::from_chars(text.data(), text.data() + text.size(), value);
  return error == std::errc{} && end == text.data() + text.size() && value > 0;
}

bool parse_mode(std::string_view text, RequestedMode& mode) {
  const auto separator = text.find('x');
  if (separator == std::string_view::npos || separator == 0) return false;
  const auto refresh_separator = text.find('@', separator + 1);
  const auto height_end = refresh_separator == std::string_view::npos
                              ? text.size()
                              : refresh_separator;
  if (!parse_dimension(text.substr(0, separator), mode.width) ||
      !parse_dimension(text.substr(separator + 1,
                                   height_end - separator - 1),
                       mode.height))
    return false;
  if (refresh_separator == std::string_view::npos) return true;
  std::uint32_t refresh = 0;
  if (!parse_dimension(text.substr(refresh_separator + 1), refresh))
    return false;
  mode.refresh_millihz = refresh;
  return true;
}

bool parse_headless_output(std::string_view text,
                           headless::OutputRequest& request) {
  const auto separator = text.find(':');
  const auto name_end =
      separator == std::string_view::npos ? text.size() : separator;
  const auto name = text.substr(0, name_end);
  if (!headless::valid_output_name(name))
    return false;
  request.name = std::string(name);
  if (separator == std::string_view::npos)
    return true;

  RequestedMode mode;
  if (!parse_mode(text.substr(separator + 1), mode))
    return false;
  const auto pixels = static_cast<std::uint64_t>(mode.width) * mode.height;
  if (mode.width > output::kMaximumPhysicalExtent ||
      mode.height > output::kMaximumPhysicalExtent ||
      pixels > output::kMaximumPhysicalPixels)
    return false;
  request.width = mode.width;
  request.height = mode.height;
  request.refresh_millihertz =
      mode.refresh_millihz.value_or(headless::kDefaultOutputRefreshMillihertz);
  return true;
}

bool parse_headless_vrr(std::string_view text,
                        headless::VrrSimulationRequest &request) {
  const auto equals = text.find('=');
  if (equals == std::string_view::npos || equals == 0 ||
      !headless::valid_output_name(text.substr(0, equals)))
    return false;
  const auto range = text.substr(equals + 1);
  const auto dash = range.find('-');
  if (dash == std::string_view::npos || dash == 0 ||
      range.find('-', dash + 1) != std::string_view::npos)
    return false;
  std::uint32_t minimum = 0;
  std::uint32_t maximum = 0;
  if (!parse_dimension(range.substr(0, dash), minimum) ||
      !parse_dimension(range.substr(dash + 1), maximum) || minimum >= maximum)
    return false;
  request = {std::string(text.substr(0, equals)), minimum, maximum};
  return true;
}

bool take_optional_path(int argc, char** argv, int& index,
                        std::optional<std::string>& destination,
                        std::string_view option, std::ostream& error) {
  std::string path;
  if (!take_path(argc, argv, index, path, option, error)) return false;
  destination = std::move(path);
  return true;
}

bool validate_backend_options(const Options& options, std::ostream& error) {
  if (options.vrr_report) {
    std::error_code path_error;
    const auto exists = std::filesystem::exists(*options.vrr_report, path_error);
    if (path_error) {
      error << "gwcomp: --vrr-report path cannot be inspected\n";
      return false;
    }
    if (exists) {
      error << "gwcomp: --vrr-report path must not already exist\n";
      return false;
    }
  }
  if (options.drm_report && options.vrr_report) {
    std::error_code drm_path_error;
    std::error_code vrr_path_error;
    const auto drm_path =
        std::filesystem::absolute(*options.drm_report, drm_path_error)
            .lexically_normal();
    const auto vrr_path =
        std::filesystem::absolute(*options.vrr_report, vrr_path_error)
            .lexically_normal();
    if (drm_path_error || vrr_path_error) {
      error << "gwcomp: report paths cannot be resolved\n";
      return false;
    }
    if (drm_path == vrr_path) {
      error << "gwcomp: --drm-report and --vrr-report require distinct paths\n";
      return false;
    }
  }
  if (options.backend == Backend::Headless && options.ipc_socket.empty() &&
      options.dump_dir.empty()) {
    error << "gwcomp: --ipc-socket and --dump-dir are required\n";
    return false;
  }
  if (options.ipc_socket.empty()) {
    error << "gwcomp: --ipc-socket is required\n";
    return false;
  }
  if (options.backend == Backend::Headless) {
    if (options.dump_dir.empty()) {
      error << "gwcomp: --dump-dir is required for the headless backend\n";
      return false;
    }
    if (options.drm_device || options.drm_fd || options.external_session ||
        options.tty || options.connector || options.mode ||
        options.drm_api != DrmApiMode::Auto || options.mirror_dump_dir ||
        options.mirror_dump_trigger || options.drm_report) {
      error << "gwcomp: DRM options require --backend drm\n";
      return false;
    }
    for (const auto &vrr : options.headless_vrr) {
      const auto configured = std::ranges::find_if(
          options.headless_outputs, [&vrr](const auto &output) {
            return output.name == vrr.name;
          });
      const auto historical = options.headless_outputs.empty() &&
                              vrr.name == headless::kDefaultOutputName;
      if (configured == options.headless_outputs.end() && !historical) {
        error << "gwcomp: --headless-vrr names an unknown headless output\n";
        return false;
      }
      const auto nominal = historical
                               ? headless::kDefaultOutputRefreshMillihertz
                               : configured->refresh_millihertz;
      if (vrr.maximum_refresh_millihertz > nominal) {
        error << "gwcomp: --headless-vrr maximum must not exceed nominal "
                 "output refresh\n";
        return false;
      }
    }
    return true;
  }
  if (!options.headless_outputs.empty() || !options.headless_vrr.empty()) {
    error << "gwcomp: --headless-output and --headless-vrr require --backend "
             "headless\n";
    return false;
  }
  if (!options.dump_dir.empty()) {
    error << "gwcomp: --dump-dir is headless-only; use --mirror-dump-dir for DRM\n";
    return false;
  }
  if (options.mirror_dump_trigger && !options.mirror_dump_dir) {
    error << "gwcomp: --mirror-dump-trigger requires --mirror-dump-dir\n";
    return false;
  }
  if (options.drm_device && options.drm_fd) {
    error << "gwcomp: --drm-device and --drm-fd are mutually exclusive\n";
    return false;
  }
  if (!options.drm_device && !options.drm_fd) {
    error << "gwcomp: DRM requires --drm-device or --drm-fd\n";
    return false;
  }
  if (options.drm_fd) {
    if (!options.external_session || options.tty) {
      error << "gwcomp: --drm-fd requires --external-session and forbids --tty\n";
      return false;
    }
  } else if (options.external_session || !options.tty) {
    error << "gwcomp: direct DRM requires --tty and forbids --external-session\n";
    return false;
  }
  return true;
}

}  // namespace

ParseOptionsResult parse_options(int argc, char** argv, Options& options,
                                 std::ostream& output, std::ostream& error) {
  for (int index = 1; index < argc; ++index) {
    const std::string_view argument(argv[index]);
    if (argument == "--help") {
      print_usage(output);
      return ParseOptionsResult::ExitSuccess;
    }
    if (argument == "--version") {
      output << "gwcomp " << GW_PROJECT_VERSION << '\n';
      return ParseOptionsResult::ExitSuccess;
    }
    if (argument == "--backend") {
      std::string value;
      if (!take_path(argc, argv, index, value, argument, error))
        return ParseOptionsResult::ExitFailure;
      if (value == "headless") options.backend = Backend::Headless;
      else if (value == "drm") options.backend = Backend::Drm;
      else {
        error << "gwcomp: --backend requires headless or drm\n";
        return ParseOptionsResult::ExitFailure;
      }
      continue;
    }
    if (argument == "--ipc-socket") {
      if (!take_path(argc, argv, index, options.ipc_socket, argument, error))
        return ParseOptionsResult::ExitFailure;
      continue;
    }
    if (argument == "--dump-dir") {
      if (!take_path(argc, argv, index, options.dump_dir, argument, error))
        return ParseOptionsResult::ExitFailure;
      continue;
    }
    if (argument == "--headless-output") {
      std::string value;
      if (!take_path(argc, argv, index, value, argument, error))
        return ParseOptionsResult::ExitFailure;
      headless::OutputRequest request;
      if (!parse_headless_output(value, request)) {
        error << "gwcomp: --headless-output requires "
                 "NAME[:WIDTHxHEIGHT[@MILLIHZ]]; NAME is a 1-63 byte ASCII "
                 "identifier and dimensions are bounded to 4096x4096\n";
        return ParseOptionsResult::ExitFailure;
      }
      if (options.headless_outputs.size() >= output::kMaximumOutputs) {
        error << "gwcomp: --headless-output may be specified at most 8 times\n";
        return ParseOptionsResult::ExitFailure;
      }
      if (std::ranges::any_of(options.headless_outputs,
                              [&request](const auto& existing) {
                                return existing.name == request.name;
                              })) {
        error << "gwcomp: --headless-output names must be unique\n";
        return ParseOptionsResult::ExitFailure;
      }
      options.headless_outputs.push_back(std::move(request));
      continue;
    }
    if (argument == "--headless-vrr") {
      std::string value;
      if (!take_path(argc, argv, index, value, argument, error))
        return ParseOptionsResult::ExitFailure;
      headless::VrrSimulationRequest request;
      if (!parse_headless_vrr(value, request)) {
        error << "gwcomp: --headless-vrr requires "
                 "NAME=MIN-MILLIHZ-MAX-MILLIHZ with 0 < MIN < MAX\n";
        return ParseOptionsResult::ExitFailure;
      }
      if (std::ranges::any_of(options.headless_vrr,
                              [&request](const auto &existing) {
                                return existing.name == request.name;
                              })) {
        error << "gwcomp: --headless-vrr names must be unique\n";
        return ParseOptionsResult::ExitFailure;
      }
      options.headless_vrr.push_back(std::move(request));
      continue;
    }
    if (argument == "--drm-device") {
      if (!take_optional_path(argc, argv, index, options.drm_device, argument,
                              error))
        return ParseOptionsResult::ExitFailure;
      continue;
    }
    if (argument == "--drm-fd") {
      int fd = -1;
      if (++index >= argc || !parse_nonnegative_fd(argv[index], fd)) {
        error << "gwcomp: --drm-fd requires a non-negative integer\n";
        return ParseOptionsResult::ExitFailure;
      }
      options.drm_fd = fd;
      continue;
    }
    if (argument == "--external-session") {
      options.external_session = true;
      continue;
    }
    if (argument == "--tty") {
      if (!take_optional_path(argc, argv, index, options.tty, argument, error))
        return ParseOptionsResult::ExitFailure;
      continue;
    }
    if (argument == "--connector") {
      if (!take_optional_path(argc, argv, index, options.connector, argument,
                              error))
        return ParseOptionsResult::ExitFailure;
      continue;
    }
    if (argument == "--mode") {
      RequestedMode mode;
      if (++index >= argc || !parse_mode(argv[index], mode)) {
        error << "gwcomp: --mode requires WIDTHxHEIGHT[@MILLIHZ]\n";
        return ParseOptionsResult::ExitFailure;
      }
      options.mode = mode;
      continue;
    }
    if (argument == "--drm-api") {
      std::string value;
      if (!take_path(argc, argv, index, value, argument, error))
        return ParseOptionsResult::ExitFailure;
      if (value == "auto") options.drm_api = DrmApiMode::Auto;
      else if (value == "atomic") options.drm_api = DrmApiMode::Atomic;
      else if (value == "legacy") options.drm_api = DrmApiMode::Legacy;
      else {
        error << "gwcomp: --drm-api requires auto, atomic, or legacy\n";
        return ParseOptionsResult::ExitFailure;
      }
      continue;
    }
    if (argument == "--mirror-dump-dir") {
      if (!take_optional_path(argc, argv, index, options.mirror_dump_dir,
                              argument, error))
        return ParseOptionsResult::ExitFailure;
      continue;
    }
    if (argument == "--mirror-dump-trigger") {
      if (!take_optional_path(argc, argv, index, options.mirror_dump_trigger,
                              argument, error))
        return ParseOptionsResult::ExitFailure;
      continue;
    }
    if (argument == "--drm-report") {
      if (!take_optional_path(argc, argv, index, options.drm_report, argument,
                              error))
        return ParseOptionsResult::ExitFailure;
      continue;
    }
    if (argument == "--vrr-report") {
      if (!take_optional_path(argc, argv, index, options.vrr_report, argument,
                              error))
        return ParseOptionsResult::ExitFailure;
      continue;
    }
    if (argument == "--scene-manifest") {
      std::string path;
      if (!take_path(argc, argv, index, path, argument, error))
        return ParseOptionsResult::ExitFailure;
      options.scene_manifest = path;
      continue;
    }
    if (argument == "--renderer") {
      std::string value;
      if (!take_path(argc, argv, index, value, argument, error))
        return ParseOptionsResult::ExitFailure;
      if (value == "software")
        options.renderer = gw::render::RendererRequest::Software;
      else if (value == "gles")
        options.renderer = gw::render::RendererRequest::Gles;
      else if (value == "auto")
        options.renderer = gw::render::RendererRequest::Auto;
      else {
        error << "gwcomp: --renderer requires software, gles, or auto\n";
        return ParseOptionsResult::ExitFailure;
      }
      continue;
    }
    if (argument == "--renderer-report") {
      if (!take_optional_path(argc, argv, index, options.renderer_report,
                              argument, error))
        return ParseOptionsResult::ExitFailure;
      continue;
    }
    if (argument == "--once") {
      options.once = true;
      continue;
    }
    if (argument == "--max-frames") {
      std::uint64_t count = 0;
      if (++index >= argc || !parse_positive(argv[index], count)) {
        error << "gwcomp: --max-frames requires a positive integer\n";
        return ParseOptionsResult::ExitFailure;
      }
      options.max_frames = count;
      continue;
    }
    error << "gwcomp: unknown option: " << argument << '\n';
    print_usage(error);
    return ParseOptionsResult::ExitFailure;
  }

  if (!validate_backend_options(options, error)) {
    print_usage(error);
    return ParseOptionsResult::ExitFailure;
  }
  return ParseOptionsResult::Run;
}

}  // namespace glasswyrm::compositor

File: crates/gwcomp/src/options.rs
use std::collections::BTreeSet;
use std::path::PathBuf;

pub const USAGE: &str = "Usage: gwcomp [--backend headless] --ipc-socket PATH\n  headless: [--dump-dir PATH] [--headless-output NAME[:WIDTHxHEIGHT[@MILLIHZ]]]...\n            [--headless-vrr NAME=MIN-MILLIHZ-MAX-MILLIHZ]...\n  renderer: [--renderer software|auto]\n  evidence: [--scene-manifest PATH] [--vrr-report PATH]\n  common: [--once] [--max-frames N] [--help] [--version]\n";

const MAXIMUM_OUTPUTS: usize = 8;
const MAXIMUM_EXTENT: u32 = 4096;
const MAXIMUM_PIXELS: u64 = 16_777_216;
const MAXIMUM_TOTAL_PIXELS: u64 = 67_108_864;

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct HeadlessOutput {
    pub name: String,
    pub width: u32,
    pub height: u32,
    pub refresh_millihertz: u32,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct HeadlessVrr {
    pub name: String,
    pub minimum_refresh_millihertz: u32,
    pub maximum_refresh_millihertz: u32,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct Options {
    pub ipc_socket: PathBuf,
    pub dump_dir: Option<PathBuf>,
    pub outputs: Vec<HeadlessOutput>,
    pub vrr: Vec<HeadlessVrr>,
    pub scene_manifest: Option<PathBuf>,
    pub vrr_report: Option<PathBuf>,
    pub once: bool,
    pub max_frames: Option<u64>,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub enum ParseOutcome {
    Run(Options),
    ExitSuccess(String),
}

/// Parses the bounded Rust transition compositor command line.
///
/// # Errors
///
/// Returns a diagnostic when an option is unknown, lacks a value, requests an
/// unavailable hardware path, or describes an invalid headless topology.
#[allow(clippy::too_many_lines)]
pub fn parse_options(arguments: impl IntoIterator<Item = String>) -> Result<ParseOutcome, String> {
    let mut arguments = arguments.into_iter();
    let _program = arguments.next();
    let mut socket = None;
    let mut dump = None;
    let mut outputs = Vec::new();
    let mut vrr = Vec::new();
    let mut scene_manifest = None;
    let mut vrr_report = None;
    let mut once = false;
    let mut max_frames = None;
    while let Some(argument) = arguments.next() {
        match argument.as_str() {
            "--help" => return Ok(ParseOutcome::ExitSuccess(USAGE.to_owned())),
            "--version" => {
                return Ok(ParseOutcome::ExitSuccess(format!(
                    "gwcomp {}\n",
                    env!("CARGO_PKG_VERSION")
                )));
            }
            "--backend" => {
                let value = take_value(&mut arguments, "--backend")?;
                if value != "headless" {
                    return Err("--backend requires headless; Rust DRM is not implemented".into());
                }
            }
            "--ipc-socket" => {
                socket = Some(PathBuf::from(take_value(&mut arguments, "--ipc-socket")?));
            }
            "--dump-dir" => dump = Some(PathBuf::from(take_value(&mut arguments, "--dump-dir")?)),
            "--headless-output" => {
                if outputs.len() == MAXIMUM_OUTPUTS {
                    return Err("--headless-output may be specified at most 8 times".into());
                }
                outputs.push(parse_output(&take_value(
                    &mut arguments,
                    "--headless-output",
                )?)?);
            }
            "--headless-vrr" => {
                vrr.push(parse_vrr(&take_value(&mut arguments, "--headless-vrr")?)?);
            }
            "--renderer" => {
                let value = take_value(&mut arguments, "--renderer")?;
                if value != "software" && value != "auto" {
                    return Err(
                        "--renderer requires software or auto; GLES is not implemented".into(),
                    );
                }
            }
            "--scene-manifest" => {
                scene_manifest = Some(PathBuf::from(take_value(
                    &mut arguments,
                    "--scene-manifest",
                )?));
            }
            "--vrr-report" => {
                vrr_report = Some(PathBuf::from(take_value(&mut arguments, "--vrr-report")?));
            }
            "--once" => once = true,
            "--max-frames" => {
                let value = take_value(&mut arguments, "--max-frames")?;
                max_frames = Some(
                    value
                        .parse::<u64>()
                        .ok()
                        .filter(|value| *value != 0)
                        .ok_or_else(|| "--max-frames requires a positive integer".to_owned())?,
                );
            }
            value
                if value.starts_with("--drm")
                    || matches!(
                        value,
                        "--tty"
                            | "--external-session"
                            | "--connector"
                            | "--mode"
                            | "--mirror-dump-dir"
                            | "--mirror-dump-trigger"
                    ) =>
            {
                return Err(
                    "DRM options are unavailable in the Rust headless transition process".into(),
                );
            }
            _ => return Err(format!("unknown option: {argument}")),
        }
    }
    let ipc_socket = socket.ok_or_else(|| "--ipc-socket is required".to_owned())?;
    if outputs.is_empty() {
        outputs.push(HeadlessOutput {
            name: "HEADLESS-1".into(),
            width: 1024,
            height: 768,
            refresh_millihertz: 60_000,
        });
    }
    let output_names: BTreeSet<_> = outputs.iter().map(|output| output.name.as_str()).collect();
    if output_names.len() != outputs.len() {
        return Err("--headless-output names must be unique".into());
    }
    let total_pixels = outputs.iter().try_fold(0_u64, |total, output| {
        total.checked_add(u64::from(output.width) * u64::from(output.height))
    });
    if total_pixels.is_none_or(|pixels| pixels > MAXIMUM_TOTAL_PIXELS) {
        return Err("headless topology exceeds 67108864 total physical pixels".into());
    }
    let vrr_names: BTreeSet<_> = vrr.iter().map(|request| request.name.as_str()).collect();
    if vrr_names.len() != vrr.len() {
        return Err("--headless-vrr names must be unique".into());
    }
    for request in &vrr {
        let output = outputs
            .iter()
            .find(|output| output.name == request.name)
            .ok_or_else(|| "--headless-vrr names an unknown headless output".to_owned())?;
        if request.maximum_refresh_millihertz > output.refresh_millihertz {
            return Err("--headless-vrr maximum must not exceed nominal output refresh".into());
        }
    }
    Ok(ParseOutcome::Run(Options {
        ipc_socket,
        dump_dir: dump,
        outputs,
        vrr,
        scene_manifest,
        vrr_report,
        once,
        max_frames,
    }))
}

fn take_value(
    arguments: &mut impl Iterator<Item = String>,
    option: &str,
) -> Result<String, String> {
    arguments
        .next()
        .filter(|value| !value.is_empty())
        .ok_or_else(|| format!("{option} requires a non-empty value"))
}

fn valid_name(name: &str) -> bool {
    (1..=63).contains(&name.len())
        && name
            .as_bytes()
            .first()
            .is_some_and(u8::is_ascii_alphanumeric)
        && name
            .bytes()
            .all(|byte| byte.is_ascii_alphanumeric() || matches!(byte, b'-' | b'_' | b'.'))
}

fn parse_output(value: &str) -> Result<HeadlessOutput, String> {
    let (name, mode) = value
        .split_once(':')
        .map_or((value, None), |(name, mode)| (name, Some(mode)));
    if !valid_name(name) {
        return Err("--headless-output requires a 1-63 byte ASCII identifier".into());
    }
    let (width, height, refresh) = if let Some(mode) = mode {
        let (extent, refresh) = mode
            .split_once('@')
            .map_or((mode, 60_000), |(extent, refresh)| {
                (extent, refresh.parse().unwrap_or(0))
            });
        let (width, height) = extent
            .split_once('x')
            .ok_or_else(|| "--headless-output requires NAME[:WIDTHxHEIGHT[@MILLIHZ]]".to_owned())?;
        (
            width.parse().unwrap_or(0),
            height.parse().unwrap_or(0),
            refresh,
        )
    } else {
        (1024, 768, 60_000)
    };
    let pixels = u64::from(width) * u64::from(height);
    if width == 0
        || height == 0
        || refresh == 0
        || width > MAXIMUM_EXTENT
        || height > MAXIMUM_EXTENT
        || pixels > MAXIMUM_PIXELS
    {
        return Err(
            "--headless-output dimensions must be within 1..4096 and at most 16777216 pixels"
                .into(),
        );
    }
    Ok(HeadlessOutput {
        name: name.into(),
        width,
        height,
        refresh_millihertz: refresh,
    })
}

fn parse_vrr(value: &str) -> Result<HeadlessVrr, String> {
    let (name, range) = value
        .split_once('=')
        .ok_or_else(|| "--headless-vrr requires NAME=MIN-MILLIHZ-MAX-MILLIHZ".to_owned())?;
    let (minimum, maximum) = range
        .split_once('-')
        .ok_or_else(|| "--headless-vrr requires NAME=MIN-MILLIHZ-MAX-MILLIHZ".to_owned())?;
    let minimum = minimum.parse().unwrap_or(0);
    let maximum = maximum.parse().unwrap_or(0);
    if !valid_name(name) || minimum == 0 || minimum >= maximum || range.matches('-').count() != 1 {
        return Err(
            "--headless-vrr requires NAME=MIN-MILLIHZ-MAX-MILLIHZ with 0 < MIN < MAX".into(),
        );
    }
    Ok(HeadlessVrr {
        name: name.into(),
        minimum_refresh_millihertz: minimum,
        maximum_refresh_millihertz: maximum,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_legacy_headless_shape() {
        let result = parse_options(
            [
                "gwcomp",
                "--ipc-socket",
                "/tmp/gw.sock",
                "--dump-dir",
                "/tmp/dumps",
                "--headless-output",
                "LEFT:640x480@60000",
                "--headless-vrr",
                "LEFT=40000-60000",
            ]
            .map(str::to_owned),
        )
        .unwrap();
        let ParseOutcome::Run(options) = result else {
            panic!("expected run")
        };
        assert_eq!(options.outputs[0].width, 640);
        assert_eq!(options.vrr[0].minimum_refresh_millihertz, 40_000);
    }

    #[test]
    fn rejects_hardware_and_bad_vrr() {
        assert!(parse_options(["gwcomp", "--backend", "drm"].map(str::to_owned)).is_err());
        assert!(
            parse_options(
                [
                    "gwcomp",
                    "--ipc-socket",
                    "s",
                    "--dump-dir",
                    "d",
                    "--headless-vrr",
                    "missing=40-60"
                ]
                .map(str::to_owned)
            )
            .is_err()
        );
    }

    #[test]
    fn parses_evidence_paths_and_rejects_excessive_topology() {
        let ParseOutcome::Run(options) = parse_options(
            [
                "gwcomp",
                "--ipc-socket",
                "s",
                "--dump-dir",
                "d",
                "--scene-manifest",
                "scene.jsonl",
                "--vrr-report",
                "vrr.jsonl",
            ]
            .map(str::to_owned),
        )
        .unwrap() else {
            panic!("expected run")
        };
        assert_eq!(options.scene_manifest, Some(PathBuf::from("scene.jsonl")));
        assert_eq!(options.vrr_report, Some(PathBuf::from("vrr.jsonl")));

        let mut arguments = vec![
            "gwcomp".to_owned(),
            "--ipc-socket".to_owned(),
            "s".to_owned(),
            "--dump-dir".to_owned(),
            "d".to_owned(),
        ];
        for index in 0..5 {
            arguments.extend([
                "--headless-output".to_owned(),
                format!("H{index}:4096x4096"),
            ]);
        }
        assert!(parse_options(arguments).is_err());
    }

    #[test]
    fn dump_output_is_explicit_and_frame_stop_remains_optional() {
        let ParseOutcome::Run(options) =
            parse_options(["gwcomp", "--ipc-socket", "s"].map(str::to_owned)).unwrap()
        else {
            panic!("expected run")
        };
        assert_eq!(options.dump_dir, None);
        assert_eq!(options.max_frames, None);

        let ParseOutcome::Run(options) = parse_options(
            [
                "gwcomp",
                "--ipc-socket",
                "s",
                "--max-frames",
                "18446744073709551615",
            ]
            .map(str::to_owned),
        )
        .unwrap() else {
            panic!("expected run")
        };
        assert_eq!(options.max_frames, Some(u64::MAX));
    }
}
